blog-post/blog-post.md

```diff
@@ -0,0 +1,634 @@
+example, you cannot create a `Native` payment with a token field or an `ERC721`
+payment with an amount field.
+
+```
+data tokenid = tokenid(word);
+
+data Payment =
+    Native(address, word)
+  | ERC20(address, address, address, word)
+  | ERC721(address, address, address, tokenid);
+```
+
+The implementation of functions `processPayment` and `calculateFee` can be made
+much simpler by using pattern matching:
+
+```
+function processPayment(payment : Payment) {
+    match payment {
+    | Native(to,amount) =>
+        transfer(to,amount);
+    | ERC20(token,from,to,amount) =>
+        transferFromERC20(from, to, amount);
+    | ERC721(token,from,to,tokenId) =>
+        transferFromERC721(from, to, tokenId);
+    }
+}
+```
+
+Another effect of the use of pattern matching / algebraic data types is that,
+since it is not possible to represent invalid states, no runtime validations
+using `require` is necessary, since they are enforced by Core Solidity type system.
+
+In this short example, we can see how Core Solidity gives the developer the tools
+to write shorter and safer code. By using algebraic data types and pattern matching,
+we can avoid the manipulation of invalid states and the need of runtime validations
+using `require`. Also, parametric polymorphism (a.k.a. generics) and type classes
+open up new possibilities for the development of safer libraries for the modular
+development of smart-contracts.
+
+### Compile Time Evaluation
+
+While not yet implemented, we are planning to add a robust system for compile time evaluation. We
+expect to support at least the compile time evaluation of pure expressions that do not require
+access to memory. Some key design questions that remain open relate to the degree to which we will
+support access to memory at compile time, the need for compile time specific primitives (e.g. string
+concatenation / hashing), as well as the need for reflection.
+
+We want to make sure the needs of the community are met here. Please feel free to reach out
+The Solidity Forum, or our Matrix channel if you have concrete use cases in mind for
+this feature.
+
+We will publish more once we have a concrete design / prototype implementation.
+
+## SAIL, Desugaring and the Standard Library
+
+Desugaring is the process of translating high-level language constructs into
+a semantically equivalent form using language's more primitive features.
+Some of the Core Solidity new high-level constructs will be translated
+into a intermediate language called SAIL - Solidity Algebraic Intermediate
+Language. Anonymous functions, assignments, contracts, indexed and storage
+access are examples of high-level Core Solidity features with are translated
+into SAIL. The SAIL language is a typed programming language featuring
+generics, typeclasses, algebraic data types and pattern matchings and it
+serves as a high-level intermediate step between Core Solidity and Yul.
```

Comment:
I think this would be clearer in reverse order. I'd first say what SAIL is and emphasize that the features mentioned in the previous section are in both Core Solidity and SAIL. Only then say that the rest is sugar. Having a grasp on what isn't sugar is necessary to understand what the sugar is.

Reply:
This section has been significantly reworked, how do you find it now?